flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java

```diff
@@ -1058,6 +1092,30 @@ public void handleError(Throwable throwable) {
 		}
 	}
 
+	/**
+	 * The heartbeat listener for JobManager and ResourceManager, they can be distinguished by ResourceID
+	 * and trigger different processes.
+	 */
+	private final class JMRMHeartbeatListener implements HeartbeatListener {
+
+		JMRMHeartbeatListener() {
+		}
+
+		@Override
+		public void notifyHeartbeatTimeout(final ResourceID resourceID) {
+			log.info("Notify heartbeat timeout for resourceID {}", resourceID);
```

Comment:
I think it would be enough to create the respective handler method in the `TaskExecutor`. The actual handling can then be implemented in a separate PR. But that way, we can easily test the behaviour by checking whether this method was called in case of a timeout.